### PR TITLE
Enable the LAN7800 PHY auto-downshift feature

### PR DESCRIPTION
**Pull request by pelwell**
Ethernet cables with faulty or missing pairs (specifically pairs C and
D) allow auto-negotiation to 1000Mbs, but do not support the successful
establishment of a link. Add a DT property, "microchip,downshift-after",
to configure the number of auto-negotiation failures after which it
falls back to 100Mbs. Valid values are 2, 3, 4, 5 and 0, where 0 means
never downshift.

Also add two new DT parameters so this behaviour can be controlled by
the user.

See: https://github.com/raspberrypi/linux/issues/2475

### CODE REVIEW

File: arch/arm/boot/dts/bcm283x-rpi-lan7515.dtsi [hunk 1]
```diff
@@ -36,9 +46,11 @@
...
-		eee = <&ethernet>,"microchip,eee-enabled?";
-		tx_lpi_timer = <&ethernet>,"microchip,tx-lpi-timer:0";
-		eth_led0 = <&ethernet>,"microchip,led-modes:0";
-		eth_led1 = <&ethernet>,"microchip,led-modes:4";
+		eee = <&eth_phy>,"microchip,eee-enabled?";
+		tx_lpi_timer = <&eth_phy>,"microchip,tx-lpi-timer:0";
+		eth_led0 = <&eth_phy>,"microchip,led-modes:0";
+		eth_led1 = <&eth_phy>,"microchip,led-modes:4";
+		eth_downshift_after = <&eth_phy>,"microchip,downshift-after";
+		eth_max_speed = <&eth_phy>,"max-speed";
```
Comment (6by9) [inline arch/arm/boot/dts/bcm283x-rpi-lan7515.dtsi]: Have I missed this eth_max_speed parameter somewhere? I don't see it in the binding nor bcm283x-rpi-lan7515.dtsi

**Review by 6by9 — COMMENTED** [review]
Your commit text on "net: lan78xx: Support auto-downshift to 100Mb/s " will get comments should you try upstreaming.

10baseT and 100baseT require 2 pairs, so comments on single pair cat3 are erroneous.

10baseT and 100baseT require 2 pairs, whilst 1000baseT requires 4. Some installations have repurposed the spare pairs from 10 and 100baseT for other uses (eg passive PoE), which results in 1000baseT being negotiated but unable to fully establish.

Also in "BCM270X_DT: Add new Ethernet DT parameters" you've got a typo of "activatess" in the commit text.

Did you find cables to test this on? If not I'll grab my hacked cable from home and test tomorrow.

File: arch/arm/boot/dts/bcm283x-rpi-lan7515.dtsi (hunk 1 above)
Comment (pelwell) [reply arch/arm/boot/dts/bcm283x-rpi-lan7515.dtsi]: The eth_max_speed parameter is defined above, but the `max-speed` property it sets is a common to all(?) PHYs, hence the lack of a vendor prefix.

**Comment by pelwell** [comment]
I've fixed the pairs reference and the typo.

File: arch/arm/boot/dts/bcm283x-rpi-lan7515.dtsi (hunk 1 above)
Comment (6by9) [reply arch/arm/boot/dts/bcm283x-rpi-lan7515.dtsi]: Probably my misunderstanding then.
I thought the overrides section could only change existing parameters. I can't see max-speed having been declared in the eth-phy node, so was confused as to what it was going to be changing. Can overrides add nodes?

**Comment by 6by9** [comment]
Still a reference to cat3 in "ARM: dts: bcm283x: Set downshift-after to 2" commit text if we really care.

File: arch/arm/boot/dts/bcm283x-rpi-lan7515.dtsi (hunk 1 above)
Comment (pelwell) [reply arch/arm/boot/dts/bcm283x-rpi-lan7515.dtsi]: Overrides can add properties, and by enabling fragments they can add nodes.

Comment (6by9) [reply arch/arm/boot/dts/bcm283x-rpi-lan7515.dtsi]: Ah, I learn something new then :-)

**Comment by pelwell** [comment]
I've reworded the 2nd commit message.

**Comment by 6by9** [comment]
Yes that seems to work pretty well.

I had minor reservations as my system uses an NFS root file system, but as lan7800 is builtin to kernel7.img on the SD card it works fine. I guess full network boot is going to fail in these conditions as the bootrom won't set this up, but not much that can be done there.
NFS booting at 10Mbit/s is a touch slow :-)

Are we worrying about upstreaming? Checkpatch complains about patch 2.
```
WARNING: DT binding docs and includes should be a separate patch. See: Documentation/devicetree/bindings/submitting-patches.txt

ERROR: spaces required around that '=' (ctx:VxW)
#49: FILE: drivers/net/phy/microchip.c:265:
+		u32 val= LAN78XX_PHY_CTRL3_AUTO_DOWNSHIFT;
 		       ^

ERROR: trailing statements should be on next line
#52: FILE: drivers/net/phy/microchip.c:268:
+		case 2:	val |= LAN78XX_PHY_CTRL3_DOWNSHIFT_CTRL_2;

ERROR: trailing statements should be on next line
#54: FILE: drivers/net/phy/microchip.c:270:
+		case 3:	val |= LAN78XX_PHY_CTRL3_DOWNSHIFT_CTRL_3;

ERROR: trailing statements should be on next line
#56: FILE: drivers/net/phy/microchip.c:272:
+		case 4:	val |= LAN78XX_PHY_CTRL3_DOWNSHIFT_CTRL_4;

ERROR: trailing statements should be on next line
#58: FILE: drivers/net/phy/microchip.c:274:
+		case 5:	val |= LAN78XX_PHY_CTRL3_DOWNSHIFT_CTRL_5;

total: 5 errors, 1 warnings, 0 checks, 62 lines checked
```
And patch 3 still references cat3.
I'm not fussed, so up to you whether you amend or not. Functionally it's good to go.

**Comment by pelwell** [comment]
Patched up and merged - thanks for the review.

**Comment by ghost** [comment]
@pelwell you might want to edit the original description on this bug, since AIUI the problem this patch was developed for is when the cable only has 2 pairs. This is not the same as a cat 3 cable. Just for super-completeness so there's no potential for misunderstanding in future.

**Comment by pelwell** [comment]
Hmm - where have I heard that before? ;-)

Updated.

**Pull request merged**